### PR TITLE
Podman Go bindings blog

### PR DESCRIPTION
**Pull request by lsm5**
Signed-off-by: Lokesh Mandvekar <lsm5@fedoraproject.org>

@ParkerVR PTAL mainly RE: your info in _config.yml

### CODE REVIEW

File: _posts/2020-08-10-podman-go-bindings.md [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+
+### Pull an image <a name="pull-images"></a>
+
+Next, we will pull a couple of images using the images.Pull() binding.
+This binding takes three arguments:
+    - The context variable created by the bindings.NewConnection() call in the first example
+    - The image name
+    - Options for image pull
+
+Append the following lines to your function.
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
**Append the following lines to your function:**
```

File: _posts/2020-08-10-podman-go-bindings.md [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+     	rawImage := "registry.fedoraproject.org/fedora:latest"
+        fmt.Println("Pulling Fedora image...")
+        _, err = images.Pull(connText, rawImage, entities.ImagePullOptions{})
+        if err != nil {
+                fmt.Println(err)
+                os.Exit(1)
+        }
+```
+
+Run it:
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
**Run it:**
```

File: _posts/2020-08-10-podman-go-bindings.md [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+
+
+### List images <a name="list-images"></a>
+Next, we will pull an image using the images.List() binding.
+This binding takes three arguments:
+   - The context variable created earlier
+   - An optional bool 'all' 
+   - An optional map of filters
+
+Append the following lines to your function.
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
**Append the following lines to your function:**
```

File: _posts/2020-08-10-podman-go-bindings.md [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+        }
+        var names []string
+        for _, i := range imageSummary {
+            names = append(names, i.RepoTags...)
+        }
+		fmt.Println(“Listing images…”)
+        fmt.Println(names)
+```
+
+Run it:
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
**Run it:**
```

File: _posts/2020-08-10-podman-go-bindings.md [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+
+After the container is started, it's a good idea to ensure the container is
+in a running state before you proceed with further operations.
+The containers.Wait() takes care of that.
+containers.Wait() takes three arguments:
+    - the context
+    - the name or ID of the container created
+    - container state (running/paused/stopped)
+  
+Append the following lines to your function.
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
**Append the following lines to your function:**
```

File: _posts/2020-08-10-podman-go-bindings.md [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+
+        running := define.ContainerStateRunning
+        _, err = containers.Wait(connText, r.ID, &running)
+        if err != nil {
+                fmt.Println(err)
+                os.Exit(1)
+        }
+```
+
+Run it:
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
**Run it:**
```

File: _posts/2020-08-10-podman-go-bindings.md [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+containers.List() takes seven arguments:
+	- the context
+	- output filters
+	- boolean to show all containers, by default only running containers are listed
+	- number of latest created containers, all states (running/paused/stopped)
+	- boolean to print pod information
+	- boolean to print rootfs size
+	- boolean to print oci runtime and container state
+
+Append the following lines to your function.
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
**Append the following lines to your function:**  
```

File: _posts/2020-08-10-podman-go-bindings.md [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+    var latestContainers = 1
+    containerLatestList, err := containers.List(connText, nil, nil, &latestContainers, nil, nil, nil)
+    if err != nil {
+        fmt.Println(err)
+        os.Exit(1)
+    }
+    fmt.Printf("Latest container is %s\n", containerLatestList[0].Names[0])
+```
+
+Run it:
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
**Run it:**
```

File: _posts/2020-08-10-podman-go-bindings.md [hunk 9]
```diff
@@ -0,0 +1,570 @@
...
+
+### Inspect Container <a name="inspect-container"></a>
+Containers can be inspected using the containers.Inspect() binding.
+containers.Inspect() takes 3 arguments:
+    - context
+    - image name or ID
+    - optional boolean to check for container size
+
+
+Append the following lines to your function:
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
**Append the following lines to your function:**
```

File: _posts/2020-08-10-podman-go-bindings.md [hunk 10]
```diff
@@ -0,0 +1,570 @@
...
+        ctrData, err := containers.Inspect(connText, r.ID, nil)
+        if err != nil {
+                fmt.Println(err)
+                os.Exit(1)
+        }
+        fmt.Printf("Container uses image %s\n", ctrData.ImageName)
+        fmt.Printf("Container running status is %s\n", ctrData.State.Status)
+```
+
+Run it:
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
**Run it:**
```

File: _posts/2020-08-10-podman-go-bindings.md [hunk 11]
```diff
@@ -0,0 +1,570 @@
...
+
+### Stop Container <a name="stop-container"></a>
+
+A container can be stopped by the containers.Stop() binding.
+containers.Stop() takes 3 arguments:
+    - context
+    - image name or ID
+    - optional timeout
+
+Append the following lines to your function:
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
**Append the following lines to your function:**
```

File: _posts/2020-08-10-podman-go-bindings.md [hunk 12]
```diff
@@ -0,0 +1,572 @@
...
+        }
+        ctrData, err = containers.Inspect(connText, r.ID, nil)
+        if err != nil {
+                fmt.Println(err)
+                os.Exit(1)
+        }
+        fmt.Printf("Container running status is now %s\n", ctrData.State.Status)
+```
+
+Run it:
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
**Run it:**
```

File: _posts/2020-08-10-podman-go-bindings.md [hunk 13]
```diff
@@ -0,0 +1,572 @@
...
+
+## Introduction
+
+In the release of Podman 2.0, we removed the experimental tag
+from its recently introduced RESTful service. While it might
+be interesting to interact with a RESTFul server using curl,
+using a set of Go based bindings is probably a more direct
+route to a production ready application.  Let’s take a look
+at how easily that can be accomplished.
+
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md:23]: Add this on this line:
`<!--readmore-->`

**Review by TomSweeneyRedHat — COMMENTED** [review]
missing the 2020-08-10-new.md

**Comment by TomSweeneyRedHat** [comment]
and it looks like you snuck the new file in under my nose, it looks good.

File: _posts/2020-08-10-podman-go-bindings.md (hunk 13 above)
Comment (lsm5) [reply _posts/2020-08-10-podman-go-bindings.md:23]: done, PTAL

File: _posts/2020-08-10-podman-go-bindings.md [hunk 14]
```diff
@@ -0,0 +1,573 @@
...
+## Contribute
+- Any issues with the bindings can be [reported upstream](https://github.com/containers/podman/issues/new/choose).
+- Check out the [Podman community page](https://podman.io/community/) for more ways to get in touch with the community.
+
+
+## Acknowledgments
+
+- This blog post was co-authored by Parker Van Roy, currently interning at Red
+Hat for summer 2020.
```
Comment (lsm5) [inline _posts/2020-08-10-podman-go-bindings.md:569]: @ParkerVR seems like there can only be 1 name in the author field, so I credited you like ^. Cool with you?

Comment (lsm5) [reply _posts/2020-08-10-podman-go-bindings.md:569]: @TomSweeneyRedHat PTAL too

Comment (lsm5) [reply _posts/2020-08-10-podman-go-bindings.md:569]: @TomSweeneyRedHat is it possible to link the names to the items in _config.yml here?

**Comment by lsm5** [comment]
one more change ...

**Comment by lsm5** [comment]
ready from my side, removing wip label...

File: _posts/2020-08-10-podman-go-bindings.md (hunk 14 above)
Comment (TomSweeneyRedHat) [reply _posts/2020-08-10-podman-go-bindings.md:569]: I'm not sure what you're asking for @lsm5

Comment (lsm5) [reply _posts/2020-08-10-podman-go-bindings.md:569]: I meant, is it possible to use some markdown magic such that the names mentioned in the acknowledgments are clickable like for the author name? Ignore me if I'm not making any sense.

Comment (TomSweeneyRedHat) [reply _posts/2020-08-10-podman-go-bindings.md:569]: If you're trying to use the default listings with the config.yml auto settings, I'm not aware of a way to do that.

File: _posts/2020-08-10-podman-go-bindings.md [hunk 15]
```diff
@@ -0,0 +1,574 @@
...
+layout: default
+author: lsm5
+categories: [blogs]
+tags: podman, containers, v2, bindings, go
+---
+![podman logo](https://podman.io/images/podman.svg)
+
+# Podman Go bindings
+{% assign author = site.authors[page.author] %}
+## By {{ author.display_name }} [GitHub](https://github.com/{{ author.github }})
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-podman-go-bindings.md]:
```suggestion
## By {{ author.display_name }} [GitHub](https://github.com/{{ author.github }}) and Parker VanRoy
```

File: _posts/2020-08-10-new.md [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+categories: [new]
+tags: containers, podman, api, v2, go, images
+---
+{% assign author = site.authors[page.author] %}
+
+In the release of Podman 2.0, we removed the experimental tag from its recently
+introduced RESTful service. While it might be interesting to interact with a
+RESTful server using curl, using a set of Go based bindings is probably a more
+direct route to a production ready application. More details from Lokesh
+Mandvekar and Parker Van Roy in this blog
```
Comment (TomSweeneyRedHat) [inline _posts/2020-08-10-new.md]:
```suggestion
Mandvekar and Parker Van Roy in this post
```

**Comment by TomSweeneyRedHat** [comment]
@lsm5 looks good overall to me.  Will take another look Monday morn with fresh eyes.

File: _posts/2020-08-10-podman-go-bindings.md (hunk 14 above)
Comment (lsm5) [reply _posts/2020-08-10-podman-go-bindings.md:569]: ack, nvm then, I guess what exists should be fine.

File: _posts/2020-08-10-new.md [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+tags: containers, podman, api, v2, go, images
+---
+{% assign author = site.authors[page.author] %}
+
+In the release of Podman 2.0, we removed the experimental tag from its recently
+introduced RESTful service. While it might be interesting to interact with a
+RESTful server using curl, using a set of Go based bindings is probably a more
+direct route to a production ready application. More details from Lokesh
+Mandvekar and Parker Van Roy in this
+[post](https://podman.io/blogs/2020/08/10/podman-go-bindings.html).
```
Comment (lsm5) [inline _posts/2020-08-10-new.md:15]: @TomSweeneyRedHat post word already exists here, i removed it from the previous line.

File: _posts/2020-08-10-podman-go-bindings.md [hunk 18]
```diff
@@ -0,0 +1,574 @@
...
+```Go
+        // Pull Busybox image (Sample 1)
+        fmt.Println("Pulling Busybox image...")
+        _, err = images.Pull(connText, "docker.io/busybox", entities.ImagePullOptions{})
+        if err != nil {
+                fmt.Println(err)
+                os.Exit(1)
+        }
+
+	    // Pull Fedora image (Sample 2)
```
Comment (rhatdan) [inline _posts/2020-08-10-podman-go-bindings.md]: Fix spacing.

File: _posts/2020-08-10-podman-go-bindings.md [hunk 19]
```diff
@@ -0,0 +1,574 @@
...
+     	rawImage := "registry.fedoraproject.org/fedora:latest"
```
Comment (rhatdan) [inline _posts/2020-08-10-podman-go-bindings.md]: Fix spacing.

File: _posts/2020-08-10-podman-go-bindings.md [hunk 20]
```diff
@@ -0,0 +1,574 @@
...
+Next, we will pull an image using the images.List() binding.
+This binding takes three arguments:
+   - The context variable created earlier
+   - An optional bool 'all' 
+   - An optional map of filters
+
+**Append the following lines to your function:**
+
+```Go
+   		// List images
```
Comment (rhatdan) [inline _posts/2020-08-10-podman-go-bindings.md]: Fix spacing.

File: _posts/2020-08-10-podman-go-bindings.md [hunk 21]
```diff
@@ -0,0 +1,574 @@
...
+        imageSummary, err := images.List(connText, nil, nil)
+        if err != nil {
+            fmt.Println(err)
+            os.Exit(1)
+        }
+        var names []string
+        for _, i := range imageSummary {
+            names = append(names, i.RepoTags...)
+        }
+		fmt.Println(“Listing images…”)
```
Comment (rhatdan) [inline _posts/2020-08-10-podman-go-bindings.md]: Fix spacing.

File: _posts/2020-08-10-podman-go-bindings.md [hunk 22]
```diff
@@ -0,0 +1,574 @@
...
+        fmt.Println(names)
```
Comment (rhatdan) [inline _posts/2020-08-10-podman-go-bindings.md:256]: Fix spacing.

**Review by rhatdan — APPROVED** [review]
LGTM other then some spacing in the code examples.

**Comment by rhatdan** [comment]
Also the author page is screwed up.

**Comment by lsm5** [comment]
> Also the author page is screwed up.

got an exact pointer please? do you mean the _config.yml?

**Comment by rhatdan** [comment]
@lsm5 When I try to view image on via github, it looks screwed up.
https://github.com/containers/podman.io/blob/7366d56dfc6487118cdf9379165160ff4c0467cd/_posts/2020-08-10-podman-go-bindings.md

**Comment by TomSweeneyRedHat** [comment]
@rhatdan @lsm5 the Author thing is displaying as expected in GitHub.  Yeah, it's way wonky there.  One it gets merged, jekyll will handle it appropriately.  I think it caught your eye this time as Lokesh added Parker's name to the end and it made it look like it only partially resolved.

**Comment by lsm5** [comment]
@TomSweeneyRedHat if Parker's mention works fine in the author section, I can remove that entry from the acknowledgments. Or the other way around if that's safer.

**Comment by rhatdan** [comment]
LGTM

**Comment by rhatdan** [comment]
This should be published on Monday or Tuesday.

**Comment by TomSweeneyRedHat** [comment]
@lsm5, this should be gtg, in fact, it here goes.

**Pull request merged**